pinot-core/src/main/java/org/apache/pinot/core/operator/transform/TransformBlockDataFetcher.java

```diff
@@ -0,0 +1,348 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.operator.transform;
+
+import java.io.Serializable;
+import org.apache.pinot.common.data.FieldSpec;
+import org.apache.pinot.common.data.FieldSpec.DataType;
+import org.apache.pinot.common.request.transform.TransformExpressionTree;
+import org.apache.pinot.common.utils.primitive.ByteArray;
+import org.apache.pinot.core.common.BlockValSet;
+import org.apache.pinot.core.segment.index.readers.Dictionary;
+
+public class TransformBlockDataFetcher {
```

Comment:
We probably need a TransformBlockCache as well, to optimize for cases where the same expression is repeated multiple times in the query (so that we don't unnecessarily compute the same expression multiple times). Can be a TODO.

Reply:
there are multiple ways to handle repeated expression. Ideally, it should be using AS. we can address those things when we start supporting AS

Reply:
+1 to add cache. Typically people will do select udf(col_a) from myTable  order by udf(col_a) or select trans_udf(col_a), agg(col_b) from myTable group by trans_udf(col_a), col_b prefer by trans_udf(col_a).

Reply:
not sure if we need a cache for that or its just a matter of deduping the expressions before they are sent to transform operator. Anyhow, we should address these optimizations once we move to the new query format.